cdc/sink/codec/canal_flat.go

```diff
@@ -362,3 +441,146 @@ func (c *CanalFlatEventBatchEncoder) SetParams(params map[string]string) error {
+	}
+	return b.msg.Type, true, nil
+}
+
+// NextRowChangedEvent implements the EventBatchDecoder interface
+// `HasNext` should be called before this.
+func (b *CanalFlatEventBatchDecoder) NextRowChangedEvent() (*model.RowChangedEvent, error) {
+	if b.msg == nil || b.msg.Type != model.MqMessageTypeRow {
+		return nil, cerrors.ErrCanalDecodeFailed.GenWithStack("not found row changed event message")
+	}
+
+	var data canalFlatMessageInterface = &canalFlatMessage{}
+	if b.enableTiDBExtension {
+		data = &canalFlatMessageWithTiDBExtension{canalFlatMessage: &canalFlatMessage{}, Extensions: &tidbExtension{}}
+	}
+
+	if err := json.Unmarshal(b.msg.Value, data); err != nil {
+		return nil, errors.Trace(err)
+	}
+	b.msg = nil
+	return canalFlatMessage2RowChangedEvent(data), nil
+}
+
+// NextDDLEvent implements the EventBatchDecoder interface
+// `HasNext` should be called before this.
+func (b *CanalFlatEventBatchDecoder) NextDDLEvent() (*model.DDLEvent, error) {
+	if b.msg == nil || b.msg.Type != model.MqMessageTypeDDL {
+		return nil, cerrors.ErrCanalDecodeFailed.GenWithStack("not found ddl event message")
+	}
+
+	var data canalFlatMessageInterface = &canalFlatMessage{}
+	if b.enableTiDBExtension {
+		data = &canalFlatMessageWithTiDBExtension{canalFlatMessage: &canalFlatMessage{}, Extensions: &tidbExtension{}}
+	}
+
+	if err := json.Unmarshal(b.msg.Value, data); err != nil {
+		return nil, errors.Trace(err)
+	}
+	b.msg = nil
+	return canalFlatMessage2DDLEvent(data), nil
+}
+
+// NextResolvedEvent implements the EventBatchDecoder interface
+// `HasNext` should be called before this.
+func (b *CanalFlatEventBatchDecoder) NextResolvedEvent() (uint64, error) {
+	if b.msg == nil || b.msg.Type != model.MqMessageTypeResolved {
+		return 0, cerrors.ErrCanalDecodeFailed.GenWithStack("not found resolved event message")
+	}
+
+	message := &canalFlatMessageWithTiDBExtension{
+		canalFlatMessage: &canalFlatMessage{},
+	}
+	if err := json.Unmarshal(b.msg.Value, message); err != nil {
+		return 0, errors.Trace(err)
+	}
+	b.msg = nil
+	return message.Extensions.WatermarkTs, nil
+}
+
+func canalFlatMessage2RowChangedEvent(flatMessage canalFlatMessageInterface) *model.RowChangedEvent {
+	result := new(model.RowChangedEvent)
+	result.CommitTs = flatMessage.getCommitTs()
+	result.Table = &model.TableName{
+		Schema: *flatMessage.getSchema(),
+		Table:  *flatMessage.getTable(),
+	}
+
+	result.Columns = canalFlatJSONColumnMap2SinkColumns(flatMessage.getData(), flatMessage.getMySQLType())
+	result.PreColumns = canalFlatJSONColumnMap2SinkColumns(flatMessage.getOld(), flatMessage.getMySQLType())
+
+	return result
+}
+
+func canalFlatJSONColumnMap2SinkColumns(cols map[string]interface{}, mysqlType map[string]string) []*model.Column {
+	result := make([]*model.Column, 0, len(cols))
+	for name, value := range cols {
+		typeStr, ok := mysqlType[name]
+		if !ok {
+			log.Panic("mysql type does not found", zap.String("column name", name), zap.Any("mysqlType", mysqlType))
```

Comment:
Can this return error, instead of panic.
Supposing TiCDC really meets panic here, the replication will pause, TiCDC nodes will panic one by one (given that the table will be rescheduled.